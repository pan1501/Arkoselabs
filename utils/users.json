[
  {
    "id": 1,
    "name": "Todd",
    "favouriteAnimalId": 3,
    "status": "DISABLED",
    "token": "CVjJZnYJzqpBynnUvKPTh8wi"
  },
  {
    "id": 2,
    "name": "Mary",
    "favouriteAnimalId": 1,
    "status": "ACTIVE",
    "token": "cNCDkgUrH3jwiWymFhmjBnPy"
  },
  {
    "id": 3,
    "name": "John",
    "favouriteAnimalId": 2,
    "status": "INACTIVE",
    "token": "Boypqgrey9uDVbLURNzvtkwG"
  },
  {
    "id": 4,
    "name": "Paul",
    "favouriteAnimalId": null,
    "status": "ACTIVE",
    "token": "ZgGceHwANPqKrGyMKMzHE4im"
  },
  {
    "id": 5,
    "name": "Max",
    "favouriteAnimalId": 3,
    "status": "PENDING",
    "token": "boqRPFRGMstzRnrozEN4VRGH"
  }
]
