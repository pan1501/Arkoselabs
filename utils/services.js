import users from './users.json';
import animals from './animals.json';

export const fetchUsers = () =>
  new Promise(resolve => setTimeout(() => resolve(users), 20));

export const fetchAnimalById = id =>
  new Promise(resolve =>
    setTimeout(() => resolve(animals.find(({ id: idx }) => id === idx)), 10),
  );

export const fetchStatus = () =>
  new Promise(resolve => setTimeout(() => resolve({ ok: true }), 5));
