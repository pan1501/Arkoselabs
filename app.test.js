import users from './utils/users.json';
import * as services from './utils/services.js';
import {
  example,
  fetchUserByIdTheirFavAnimalAndStatus,
  config,
  normaliseTypes,
  omitProperties,
  flattenObject,
  getValue,
} from './app';

describe('Technical Assessment', () => {
  test('Example', () => {
    expect(example()).toBe(true);
  });

  test('Omit properties', () => {
    expect(
      omitProperties(users, ['favouriteAnimalId', 'id', 'token'])
    ).toMatchSnapshot();
  });

  test('Get config', () => {
    const getUserConfig = config({ active: true, country: 'au' });
    expect(getUserConfig({ country: 'us', color: 'blue' })).toEqual({
      active: true,
      color: 'blue',
      country: 'us',
    });
  });

  test('Fetch data', () => {
    return fetchUserByIdTheirFavAnimalAndStatus(2, services).then(
      ([animal, status, user]) => {
        expect(user.name).toEqual('Mary');
        expect(animal.name).toEqual('Cat');
        expect(status.up).toBeTruthy;
      }
    );
  });

  test('Normalise types', () => {
    const config = {
      ACTIVE: ['ACTIVE'],
      INACTIVE: ['DISABLED', 'INACTIVE', 'PENDING'],
    };

    expect(normaliseTypes(users, 'status', config)).toMatchSnapshot();
  });

  test('Flatten object', () => {
    const data = {
      user: {
        name: 'sam',
        age: 33,
        animals: {
          cats: null,
          dogs: ['spot', 'george'],
        },
      },
      account: 'web',
      loggedIn: false,
      status: {
        status: 'ok',
      },
    };

    expect(flattenObject(data)).toMatchSnapshot();
  });

  test('Get value', () => {
    const haystack = {
      id: '23fgll',
      logs: {
        rgiij: {
          logged: true,
          time: 1548385781499,
          marks: ['http', 'browser'],
        },
        rgqyy: null,
      },
      type: 'visual',
      check: ['id', 22, 23],
    };

    expect(getValue('id', haystack)).toBe('23fgll');
    expect(getValue(['logs', 'rgiij', 'marks'], haystack)).toEqual([
      'http',
      'browser',
    ]);
    expect(getValue(['logs', 'rrkoq'], haystack)).toBeNull();
  });
});
