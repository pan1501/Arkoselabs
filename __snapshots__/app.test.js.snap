// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Technical Assessment Flatten object 1`] = `
Object {
  "account": "web",
  "age": 33,
  "cats": null,
  "dogs": Array [
    "spot",
    "george",
  ],
  "loggedIn": false,
  "name": "sam",
  "status": "ok",
}
`;

exports[`Technical Assessment Normalise types 1`] = `
Array [
  Object {
    "favouriteAnimalId": 3,
    "id": 1,
    "name": "Todd",
    "status": "INACTIVE",
    "token": "CVjJZnYJzqpBynnUvKPTh8wi",
  },
  Object {
    "favouriteAnimalId": 1,
    "id": 2,
    "name": "Mary",
    "status": "ACTIVE",
    "token": "cNCDkgUrH3jwiWymFhmjBnPy",
  },
  Object {
    "favouriteAnimalId": 2,
    "id": 3,
    "name": "John",
    "status": "INACTIVE",
    "token": "Boypqgrey9uDVbLURNzvtkwG",
  },
  Object {
    "favouriteAnimalId": null,
    "id": 4,
    "name": "Paul",
    "status": "ACTIVE",
    "token": "ZgGceHwANPqKrGyMKMzHE4im",
  },
  Object {
    "favouriteAnimalId": 3,
    "id": 5,
    "name": "Max",
    "status": "INACTIVE",
    "token": "boqRPFRGMstzRnrozEN4VRGH",
  },
]
`;

exports[`Technical Assessment Omit properties 1`] = `
Array [
  Object {
    "name": "Todd",
    "status": "DISABLED",
  },
  Object {
    "name": "Mary",
    "status": "ACTIVE",
  },
  Object {
    "name": "John",
    "status": "INACTIVE",
  },
  Object {
    "name": "Paul",
    "status": "ACTIVE",
  },
  Object {
    "name": "Max",
    "status": "PENDING",
  },
]
`;
